### PR TITLE
Keyvault integration

### PR DESCRIPTION
**Pull request by sushilraje**
# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [ ] All new and existing tests passed.
- [ ] The code follows the code style and conventions of this project.
- [ ] The change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

# Description of the change
We are in the process of moving our secrets (such as conn string, resource names) from environment variables to secrets in key-vault. This PR has corresponding changes for the services to connect with KV and fetch the required env variables. This PR also contains corresponding changes to the scripts.

# Motivation for the change

...

### CODE REVIEW

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Override
+    public boolean getBool(String key) {
+        Boolean value = false;
+
+        try {
+            value = this.data.getBoolean(key);
+        } catch (ConfigException.Missing e) {
+            // Do Nothing as this goes to KV logic (below)
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: May be add log for verbose

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        } catch (ConfigException.WrongType e) {
+            // Try to get this as a String and
+            value = this.stringToBoolean(
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: log

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Override
+    public int getInt(String key) throws InvalidConfigurationException {
+        Integer value = null;
+
+        try {
+            value = this.data.getInt(key);
+        } catch (ConfigException.Missing e) {
+            // Do Nothing as this goes to KV logic (below)
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: log

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        } catch (ConfigException.WrongType e) {
+            // Try to get this as a String and
+            value = this.stringToInt(
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: log

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    Long.valueOf(this.keyVault.getKeyVaultSecret(key)),
+                    ChronoUnit.SECONDS
+            );
+        }
+
+        return value;
+    }
+
+    @Override
+    public boolean hasPath(String path) {
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: Provide some description to what this method does

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    private int stringToInt(String value, Integer defaultValue) throws InvalidConfigurationException {
+        try {
+            return Integer.valueOf(value);
+        } catch (NumberFormatException e) {
+            throw new InvalidConfigurationException("Unable to load configuration value for '{key}'", e);
+        } catch (Exception e) {
+            // If string value is not found or any other exception than NumberFormat exception.
+            return defaultValue;
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: Add logging

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            @Override
+            public String doAuthenticate(String authorization, String resource, String scope) {
+
+                AuthenticationResult authResult;
+                try {
+                    authResult = getAccessToken(authorization, resource);
+                    return authResult.getAccessToken();
+                } catch (Exception e) {
+                    // TODO: Add logging
+                    // e.printStackTrace();
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: Add logging now.

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+                }
+                return "";
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: Does empty return mean failure?

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        //Starts a service to fetch access token.
+        ExecutorService service = null;
+        try {
+            service = Executors.newFixedThreadPool(1);
+            AuthenticationContext context = new AuthenticationContext(authorization, false, service);
+
+            Future<AuthenticationResult> future = null;
+
+            //Acquires token based on client ID and client secret.
+            if (this.clientSecret != null && this.clientSecret != null) {
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: Should we check for null and empty like c#?

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                future = context.acquireToken(resource, credentials, null);
+            }
+
+            result = future.get();
+        } finally {
+            service.shutdown();
+        }
+
+        if (result == null) {
+            throw new RuntimeException("Authentication results were null.");
```
Comment (ppathan) [inline config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: log before throw?

File: config/conf/application.conf [hunk 11]
```diff
@@ -7,40 +7,40 @@
...
-    devicesimulation-webservice-url: ${?PCS_DEVICESIMULATION_WEBSERVICE_URL}
-    telemetry-webservice-url: ${?PCS_TELEMETRY_WEBSERVICE_URL}
-    seed-template: "default"
-    azuremaps-key: ${?PCS_AZUREMAPS_KEY}
+    webservicePort : ${http.port}
+    storageAdapterWebServiceUrl: ${?PCS_STORAGEADAPTER_WEBSERVICE_URL}
+    deviceSimulationWebServiceUrl: ${?PCS_DEVICESIMULATION_WEBSERVICE_URL}
+    telemetryWebServiceUrl: ${?PCS_TELEMETRY_WEBSERVICE_URL}
+    seedTemplate: "default"
+    azureMapsKey: ${?PCS_AZUREMAPS_KEY}
```
Comment (ppathan) [inline config/conf/application.conf]: In c# for lot of these static values I think we have hard coded values, if that is true should we follow the same pattern here?

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java (hunk 1 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: added

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java (hunk 3 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: added

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java (hunk 5 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java (hunk 6 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java (hunk 7 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: added

Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: This function is similar to examples from Keyvault library. It seems we need to return "" string than null value.

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java (hunk 8 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: This function is similar to examples from Keyvault library. It seems we need to return "" string than null value.

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java (hunk 9 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java (hunk 10 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java]: done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java (hunk 4 above)
Comment (sushilraje) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/ConfigData.java]: There is a log above

**Pull request merged**